tensorrt_llm/_torch/modules/fused_moe/routing.py

```diff
@@ -0,0 +1,264 @@
+            mask = (
+                (max_values - router_logits) /
+                torch.max(torch.abs(router_logits), max_values)) > 2 * self.eps
+            masked_logits = torch.where(mask, -float('inf'), router_logits)
+            softmax_masked_logits = torch.nn.functional.softmax(masked_logits,
+                                                                dim=-1)
+            selected_values = torch.gather(softmax_masked_logits, -1,
+                                           max_indices)
+            topk_values[:, i] = selected_values.squeeze(-1)
+
+        return topk_indices.to(torch.int32), topk_values
+
+
+class StaticMoeRoutingMethod(BaseMoeRoutingMethod):
+
+    def __init__(self,
+                 routing_tensor: torch.Tensor,
+                 routing_scales: Optional[torch.Tensor] = None):
+        super().__init__()
+        assert routing_tensor.dtype == torch.int32
+        if routing_scales is not None:
+            assert routing_tensor.shape[0] == routing_scales.shape[0]
+            assert routing_tensor.shape[1] == routing_scales.shape[1]
+            assert routing_scales.dtype == torch.float32
+        self.routing_tensor = routing_tensor
+        self.routing_scales = routing_scales
+
+    def apply(self,
+              router_logits: torch.Tensor) -> (torch.Tensor, torch.Tensor):
+        return self.routing_tensor, self.routing_scales
+
+    def get_experts_per_token(self):
+        return self.routing_tensor.shape[1]
+
+
+class LoadBalancedMoeRoutingMethod(BaseMoeRoutingMethod):
+
+    def __init__(self, top_k: int):
+        super().__init__()
+        self.top_k = top_k
+
+    def apply(self,
+              router_logits: torch.Tensor) -> (torch.Tensor, torch.Tensor):
+        balanced_values = torch.ones(router_logits.shape[0],
+                                     self.top_k,
+                                     device=router_logits.device,
+                                     dtype=torch.float32)
+        balanced_indices = torch.empty(router_logits.shape[0],
+                                       self.top_k,
+                                       device=router_logits.device,
+                                       dtype=torch.int32)
+
+        # Fill the balanced_indices with each expert in round-robin fashion
+        final_size = router_logits.shape[0] * self.top_k
+        repeat_count = math.ceil(final_size / router_logits.shape[1])
+        indices = torch.arange(router_logits.shape[1],
+                               device=router_logits.device,
+                               dtype=torch.int32)
+        indices = indices.repeat(repeat_count)
+        indices = indices[:final_size]
+        balanced_indices = indices.view(router_logits.shape[0],
+                                        self.top_k).contiguous()
+
+        return balanced_indices, balanced_values
+
+
+class Qwen3MoeRoutingMethod(BaseMoeRoutingMethod):
```

Comment:
Maybe I am missing something, this is just the Renormalize method spelled less efficiently

`x = topk(softmax()); x /= x.sum()` is mathematically equivalent to `softmax(topk)`

Maybe we should add a comment explaining this isomorphism so we dont get people continually re-adding this spelling

Reply:
I'll leave this to a follow-up PR. It'll need to rerun the CI to run qwen3 accuracy tests. This PR is blocking another 3 PRs. I need to get it in ASAP.